include/drivers/dai.h

```diff
@@ -0,0 +1,717 @@
+
+/**
+ * @brief Pulse density modulation.
+ *
+ * Serial interface standard used by DMICs and some codecs.
+ */
+#define DAI_FMT_DATA_FORMAT_PDM			(5 << DAI_FMT_DATA_FORMAT_SHIFT)
+
+/** Send MSB first */
+#define DAI_FMT_DATA_ORDER_MSB              (0 << 3)
+/** Send LSB first */
+#define DAI_FMT_DATA_ORDER_LSB              BIT(3)
+/** Invert bit ordering, send LSB first */
+#define DAI_FMT_DATA_ORDER_INV              DAI_FMT_DATA_ORDER_LSB
+
+/** Data Format bit field position. */
+#define DAI_FMT_CLK_FORMAT_SHIFT           4
+/** Data Format bit field mask. */
+#define DAI_FMT_CLK_FORMAT_MASK            (0x3 << DAI_FMT_CLK_FORMAT_SHIFT)
+
+/** Invert bit clock */
+#define DAI_FMT_BIT_CLK_INV                 BIT(4)
+/** Invert frame clock */
+#define DAI_FMT_FRAME_CLK_INV               BIT(5)
+
+/** NF represents "Normal Frame" whereas IF represents "Inverted Frame"
+ *  NB represents "Normal Bit Clk" whereas IB represents "Inverted Bit clk"
+ */
+#define DAI_FMT_CLK_NF_NB		(0 << DAI_FMT_CLK_FORMAT_SHIFT)
+#define DAI_FMT_CLK_NF_IB		(1 << DAI_FMT_CLK_FORMAT_SHIFT)
+#define DAI_FMT_CLK_IF_NB		(2 << DAI_FMT_CLK_FORMAT_SHIFT)
+#define DAI_FMT_CLK_IF_IB		(3 << DAI_FMT_CLK_FORMAT_SHIFT)
+
+typedef uint8_t dai_opt_t;
+
+/** Run bit clock continuously */
+#define DAI_OPT_BIT_CLK_CONT                (0 << 0)
+/** Run bit clock when sending data only */
+#define DAI_OPT_BIT_CLK_GATED               BIT(0)
+
+/** DAI driver is bit clock provider (codec is consumer) */
+#define DAI_OPT_BIT_CLK_PROVIDER              (0 << 1)
+/** DAI driver is bit clock consumer (codec is provider) */
+#define DAI_OPT_BIT_CLK_CONSUMER               BIT(1)
+/** DAI driver is frame clock provider (codec is consumer) */
+#define DAI_OPT_FRAME_CLK_PROVIDER            (0 << 2)
+/** DAI driver is frame clock consumer (codec is provider) */
+#define DAI_OPT_FRAME_CLK_CONSUMER             BIT(2)
+
+/** @brief Loop back mode.
+ *
+ * In loop back mode RX input will be connected internally to TX output.
+ * This is used primarily for testing.
+ */
+#define DAI_OPT_LOOPBACK                    BIT(7)
+
+/** @brief Ping pong mode
+ *
+ * Inherited from legacy TBD to check.
+ *
+ * In ping pong mode TX output will keep alternating between a ping buffer
+ * and a pong buffer. This is normally used in audio streams when one buffer
+ * is being populated while the other is being played (DMAed) and vice versa.
+ * So, in this mode, 2 sets of buffers fixed in size are used. Static Arrays
+ * are used to achieve this and hence they are never freed.
+ */
+#define DAI_OPT_PINGPONG                    BIT(6)
+
+/** \brief Types of DAI
+ *
+ * The type of the DAI. This ID type is used to configure bespoke DAI HW
+ * settings.
+ *
+ * DAIs have a lot of physical link feature variability and therefore need
+ * different configuration data to cater for different use cases. We
+ * sometimes may need to pass extra configuration data prior to DAI start.
+ */
+enum dai_type {
```

Comment:
Some of these enumerants look like they're runtime-discriminatory ("is this particular device a DMIC or HDA stream?") and some are clearly not (Intel vs. AMD).  I'm wondering if there's an extra level of abstraction needed here, where the SOC level code somewhere defines the enumerants that make sense for the platform being built?

Obviously not a big deal if it's a short list and rarely changes.

Reply:
Will only change when new DAI IP drivers are added to Zephyr.

Reply:
Should DAI type be stable? I.e. adding a new Intel DAI in the middle would change the value of subsequent entries - does it matter? Or do you plan to append new DAI types - this would be less readable.

Reply:
Yup, I agree this looks a bit off. @lgirdwood  proposed "dai_config_ext" and I think that sounds good. I.e. a DAI we can have an enumeration of runtime config objects (e..g filter-coefficient blobs) that drivers can support, or not support. A DAI driver that doesn't need any runtime blobs, or all configuration is static (from DT), can ignore the dai_config_ext. But for DAIs that do need this at runtime, will allow to pass the config via dai_config_ext.